src/Microsoft.AspNetCore.Razor.Language/Syntax/GreenNode.cs

```diff
@@ -0,0 +1,289 @@
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal abstract class GreenNode
+    {
+        private static readonly RazorDiagnostic[] EmptyDiagnostics = Array.Empty<RazorDiagnostic>();
+        private static readonly ConditionalWeakTable<GreenNode, RazorDiagnostic[]> diagnosticsTable =
+            new ConditionalWeakTable<GreenNode, RazorDiagnostic[]>();
+
+        private NodeFlags _flags;
+        private byte _slotCount;
+
+        protected GreenNode(SyntaxKind kind)
+        {
+            Kind = kind;
+        }
+
+        protected GreenNode(SyntaxKind kind, int fullWidth)
+            : this(kind)
+        {
+            if (fullWidth == -1)
+            {
+                throw new InvalidOperationException($"Can't create {typeof(GreenNode).Name} with {nameof(fullWidth)} {fullWidth}.");
+            }
+
+            FullWidth = fullWidth;
+        }
+
+        protected GreenNode(SyntaxKind kind, RazorDiagnostic[] diagnostics, SyntaxAnnotation[] annotations)
+            : this(kind, 0, diagnostics, annotations)
+        {
+        }
+
+        protected GreenNode(SyntaxKind kind, int fullWidth, RazorDiagnostic[] diagnostics, SyntaxAnnotation[] annotations)
+            : this(kind, fullWidth)
+        {
+            if (diagnostics?.Length > 0)
+            {
+                _flags |= NodeFlags.ContainsDiagnostics;
+                diagnosticsTable.Add(this, diagnostics);
+            }
+            if (annotations?.Length > 0)
+            {
+                foreach (var annotation in annotations)
+                    if (annotation == null)
+                        throw new ArgumentException(paramName: nameof(annotations), message: "Annotation cannot be null");
+                _flags |= NodeFlags.ContainsAnnotations;
+                //annotationsTable.Add(this, annotations);
+            }
+        }
+
+        public virtual int Width
+        {
+            get
+            {
+                return FullWidth - (GetLeadingTriviaWidth() + GetTrailingTriviaWidth());
+            }
+        }
+
+        internal virtual bool IsList => false;
+
+        internal virtual bool IsToken => false;
+
+        internal virtual bool IsMissing => (_flags & NodeFlags.IsMissing) != 0;
+
+        internal int FullWidth { get; private set; }
+
+        internal SyntaxKind Kind { get; }
+
+        protected void AdjustWidth(GreenNode node)
+        {
+            FullWidth += node == null ? 0 : node.FullWidth;
+        }
+
+        internal virtual GreenNode GetLeadingTrivia()
+        {
+            // TODO
+            return null;
+        }
+
+        public virtual int GetLeadingTriviaWidth()
+        {
+            // TODO
+            return 0;
+        }
+
+        internal virtual GreenNode GetTrailingTrivia()
+        {
+            // TODO
+            return null;
+        }
+
+        public virtual int GetTrailingTriviaWidth()
+        {
+            // TODO
+            return 0;
```

Comment:
Why are these TODOs? If this feature isn't needed right now could you add it to an issue somwhere instead? I'd really prefer not to use TODOs in code.